dali/kernels/imgproc/convolution/cutlass/threadblock/predicated_tile_iterator.h

```diff
@@ -369,20 +568,319 @@ class PredicatedTileIterator<Shape_, Element_, layout::PitchLinear, AdvanceRank,
+   * as the loads are row-wise, this can be a no-op if mask_first or mask_last is true.
+   */
+  template <bool mirrored = true, bool IsInnerConv_ = kIsInnerConv>
+  CUTLASS_DEVICE std::enable_if_t<IsInnerConv_> add_vec(AccessType &dst, int abs_window_element,
+                                                      bool mask_first, bool mask_last) {
+    static_assert(mirrored, "All accesses should be mirrored for inner conv.");
+    if (mask_first || mask_last) {
+      return;
+    }
+    auto aligned_offset = get_aligned_offset_data<mirrored>(abs_window_element);
+    constexpr int window_start = ConvWindowConfiguration::template getWindowStart<mirrored>();
+    const auto *access_ptr = reinterpret_cast<AccessType const *>(pointer_ + window_start +
+                                                                  aligned_offset.aligned_element);
+    mux_add(dst, access_ptr[0], access_ptr[1], aligned_offset.offset);
+  }
+
+  /**
+   * @brief Load and add kernel elements from abs_window_element - Outer Conv variant
+   *
+   * In inner conv, we skip first and last column for border handling,
+   * so we sometimes need to mask the addition of first of last vector element.
+   * As for "vector aligned" loads, the data must be a multpile of that size, we only do it
+   * for the first or last element.
+   */
+  template <bool mirrored, bool IsInnerConv_ = kIsInnerConv>
+  CUTLASS_DEVICE std::enable_if_t<!IsInnerConv_> add_vec(AccessType &dst, int abs_window_element,
+                                                         bool mask_first, bool mask_last) {
+    // In outer conv we skip columns
+    auto aligned_offset = get_aligned_offset_data<mirrored>(abs_window_element);
+    constexpr int window_start = ConvWindowConfiguration::template getWindowStart<mirrored>();
+    const auto *access_ptr = reinterpret_cast<AccessType const *>(pointer_ + window_start +
+                                                                  aligned_offset.aligned_element);
+    Element tmp = static_cast<Element>(0);
+
+    if (mask_first) {
+      tmp = static_cast<Element>(dst[0]);
+    } else if (mask_last) {
+      tmp = static_cast<Element>(dst[AccessSize - 1]);
+    }
+
+    mux_add(dst, access_ptr[0], access_ptr[1], aligned_offset.offset);
+
+    if (mask_first) {
+      dst[0] = static_cast<Element>(tmp);
+    } else if (mask_last) {
+      dst[AccessSize - 1] = static_cast<Element>(tmp);
+    }
+  }
+
+  /**
+   * @brief Channel information is necessary only for inner convolution, for outer it's always 1
+   */
+  CUTLASS_DEVICE
+  int Channels() {
+    if (kIsInnerConv) {
+      return channels_;
+    }
+    return 1;
+  }
+
+  /**
+   * @brief Whether to skip first row/column in border handling
+   *
+   * Note: Border reflect 1001 would always return false here
+   */
+  CUTLASS_DEVICE
+  bool mask_first(int dist_first) {
+    if (kIsInnerConv) {
+      return dist_first < Channels();
+    } else {
+      return dist_first == 0;
+    }
```

Comment:
Nitpick: wouldnt `return dist_first < Channels();` work anyway? Channels() is 1 for non-inner conv. and it doens't seem that negative dist_first is valid.

Reply:
I prefer the explicit one here.